### PR TITLE
add a vcs_auth_token to project data for interacting with private repos

### PR DESCRIPTION
**Pull request by mwrock**
This PR just covers the API back end and does not add the auth_token to any of the TS code. The field is optional so it should not break the front end.

Signed-off-by: mwrock <matt@mattwrock.com>

### CODE REVIEW

**Comment by mwrock** [comment]
Marking this WIP until I finish up end to end testing

**Comment by mwrock** [comment]
looks like we also need a `vcs_username` for the clone

**Comment by mwrock** [comment]
Testing with a private repo, POST a project to http://localhost:9636/v1/projects using jason like the following:
```
{
    "origin": "core",
    "plan_path": "components/srv-proxy/plan.sh",
    "github": {
        "organization": "habitat-sh",
        "repo": "firewatch",
        "auth_token": "<personal access token>",
        "username": "mwrock"
    }
}
```

**Comment by mwrock** [comment]
removed WIP. Verified that a private repo can be added and cloned by the worker. This is ready for review.

**Comment by mwrock** [comment]
related to #2888

File: components/builder-worker/src/vcs.rs [hunk 1]
```diff
@@ -21,24 +21,90 @@ use error::Result;
...
+                );
+                git2::Repository::clone(&self.url(), path)?;
                 Ok(())
             }
             _ => panic!("Unknown vcs type"),
         }
     }
+
+    pub fn url(&self) -> String {
+        // git urls come from the clone_url of a repo which should always
```
Comment (reset) [inline components/builder-worker/src/vcs.rs]: I think we want to play it safe here and parse the data and manipulate it that way. I would use this [crate](https://github.com/servo/rust-url) and these functions:

* [parse](https://docs.rs/url/1.5.1/url/struct.Url.html#method.parse)
* [set_username](https://docs.rs/url/1.5.1/url/struct.Url.html#method.set_username)
* [set_password](https://docs.rs/url/1.5.1/url/struct.Url.html#method.set_password)

We should validate that the "Data" in the constructor of the vcs struct before we get to this function

**Comment by mwrock** [comment]
I added validation and error handling to the url parsing and am using the url crate now @reset. I didn't put the validation in the constructor because this struct is intended to be a generic version control abstraction and the url is really git specific. That said, I do think a constructor is the right place ideally. We could make a trait out of VCS and have different versioning systems implement it. However I think its too early for that and not worth it until we have at least one other vcs.

File: components/builder-api/src/server/handlers.rs [hunk 2]
```diff
@@ -514,7 +525,7 @@ pub fn project_update(req: &mut Request) -> IronResult<Response> {
         (session_token, session_id)
     };
 
-    let (organization, repo) = match req.get::<bodyparser::Struct<ProjectCreateReq>>() {
+    let (organization, repo, token) = match req.get::<bodyparser::Struct<ProjectCreateReq>>() {
```
Comment (chefsalim) [inline components/builder-api/src/server/handlers.rs]: Might want to pull a bunch of this parse code into a central function that can be used by both project_create and project_update

Comment (mwrock) [reply components/builder-api/src/server/handlers.rs]: good call. I put all of that in `github_creds_from_body` @chefsalim

**Comment by reset** [comment]
@mwrock It looks like we've got a few conflicts from other changes in master. Once you've settled them, merge when ready!

**Comment by mwrock** [comment]
@therealpadams approve

**Comment by mwrock** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/278306583).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/278306583).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**